static_frame/core/index_level.py

```diff
@@ -174,6 +174,45 @@ def __init__(self,
             self._length = 0
 
     #---------------------------------------------------------------------------
+    def traverse(self) -> tp.Union[Index, TreeNodeT]:
+        '''Traverses the IndexLevel tree via preorder traversal
+        '''
+        if self.targets is None:
+            return self.index
+
+        tree: TreeNodeT = {}
+
+        sentinel: tp.Any = object()
+        for target, label in zip_longest(self.targets, self.index, fillvalue=sentinel):
+            assert target is not sentinel and label is not sentinel, f"Malformed IndexLevel {self}"
+            tree[label] = target.traverse()
+
+        return tree
+
+    @staticmethod
+    def pad(s: str) -> str:
+        _pad = f"{' ' * 4}"
+        return _pad + f"\n{_pad}".join(s.split("\n"))
+
+    @classmethod
+    def _format(cls, body: str, name: str = "") -> str:
+        if name:
+            name = " " + name
+        return f"{cls.__name__}:{name} {{\n{body}\n}},"
+
+    def _repr(self, name: str = "") -> str:
+        if self.targets is None:
+            return self._format(self.pad(str(self.index)), name=name)
+
+        results: tp.List[str] = []
+        for label, target in zip(self.index, self.targets):
+            results.append(target._repr(name=repr(label)))
+
+        return self._format(self.pad("\n".join(results)), name=name)
+
+    def __repr__(self) -> str:
+        return self._repr("ROOT").rstrip(',')
```

Comment:
In this version, I call it `ROOT`, but that feels a bit cheesy. Although I don't really like how it looks without a name, since all the nested `IndexLevels` have their associated name from the index of their parent IndexLevel.

Reply:
Nevermind I decided to remove it. The PR is ready to merge.